### DISCUSSION TITLE
Return removeListener from addListener

### DISCUSSION THREAD

**Discussion by pke** [react-navigation/react-navigation#9285] [Ideas]
I wonder if we could introduce a change in the return value of `Navigation.addEventLister` to return a function that removes the listener. This pattern would help working with listeners in React effect hooks

```ts
// now
React.useEffect(() => {
  function handler(event) {    
     if (someDep) ...
  }
  navigation.addListener("beforeRemove", handler)
  return () => navigation.removeListener("beforeRemove", handler)
}, [navigation, someDep])

// then
React.useEffect(() => {
  return navigation.addListener("beforeRemove", event => {
     if (someDep) ...
   })
}, [navigation, someDep])

// -- or --
React.useEffect(() => (
 navigation.addListener("beforeRemove", event => {
   if (someDep) ...
 })
), [navigation, someDep])
```

Change to the `addListener` would probably pretty easy: 

```js
function addListener(name, handler) {
  ...
  return () => removeListener(name, handler)
}

**Comment by satya164** [comment]
It already does that!

**Reply by pke** [reply]
Oh, why didn't VSCode not show that then I wonder. ;( Sorry for bothering then ;)

**Reply by pke** [reply]
@satya164 now that this is cleared up I have the strange behaviour that the "beforeRemove" listener only triggers **after**  the navigation happened. What might cause this?
